amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/common/fm/FMService.java

```diff
@@ -22,49 +22,49 @@
  * @author Nadejda Mandrescu
  */
 public class FMService {
-    protected static final Logger LOGGER = Logger.getLogger(FMService.class);
+    
+    protected static final Logger logger = Logger.getLogger(FMService.class);
     
     /**
-     * 
+     *
      * @param config
-     * @return
+     * @return fm settings
      */
-    public static Map<String, Object> getFMSettings(FMSettingsConfig config) {
-        Map<String, Object> fmSettingsResult = new HashMap<>();
+    public static FMSettings getFMSettings(FMSettingsConfig config) {
+        FMSettings fmSettings = new FMSettings();
         
-        try {
-            if (config.isValid()) {
-                if (config.getReportingFields()) {
-                    fmSettingsResult.put(EPConstants.REPORTING_FIELDS, ColumnsVisibility.getConfigurableColumns());
-                }
-                
-                if (config.getEnabledModules()) {
-                    fmSettingsResult.put(EPConstants.ENABLED_MODULES,
-                            FMSettingsMediator.getEnabledSettings(FMSettingsMediator.FMGROUP_MODULES));
-                }
-    
-                provideModulesDetails(fmSettingsResult, config);
-            } else {
-                fmSettingsResult.put(EPConstants.ERROR, "Invalid modules details requested: "
-                        + config.getDetailModules() + ". Allowed are: " + config.getAllowedModules());
+        if (config.isValid()) {
+            if (config.getReportingFields()) {
+                fmSettings.setReportingFields(ColumnsVisibility.getConfigurableColumns());
+            }
+            
+            if (config.getEnabledModules()) {
+                fmSettings.setEnabledModules(
+                        FMSettingsMediator.getEnabledSettings(FMSettingsMediator.FMGROUP_MODULES));
             }
-        } catch(Exception ex) {
-            LOGGER.error("Unexpected error occurred while generating FM settings", ex);
-            fmSettingsResult.put(EPConstants.ERROR, ex.getMessage());
+            
+            provideModulesDetails(fmSettings, config);
+        } else {
+            fmSettings.setError(String.format("Invalid modules details requested: %s. Allowed are: %s",
+                    config.getDetailModules(), config.getAllowedModules()));
         }
         
-        return fmSettingsResult;
+        return fmSettings;
     }
     
-    private static void provideModulesDetails(Map<String, Object> fmSettingsResult, FMSettingsConfig config) {
+    private static void provideModulesDetails(FMSettings fmSettings, FMSettingsConfig config) {
         List<String> detailModules = config.getDetailModules();
-        if (detailModules == null || detailModules.size() == 0) {
+        
+        FMSettingsTree settingsTree = new FMSettingsTree();
+        FMSettingsFlat settingsFlat = new FMSettingsFlat();
```

Comment:
I think it would be nicer to initialize them within `fmSettings` object and then below use `fmSettings.getFmSettings()..getModules().put(...`

Reply:
We don't know the type of T during initialization. 
Also, it means to create an interface with modules field as a generic Map. (the modules are different map in the classes)

Reply:
For parametereized class in constructor:
```
Class<T> clazz = (Class<T>) ((ParameterizedType) this.getClass().getGenericSuperclass()).getActualTypeArguments()[0];
```
And for modules we can cast when accessing fmSettings.

Anyway, was just a thought, and of course we can leave as is.

Reply:
It would be better to leave as it is ;)